chromadb/api/async_client.py

```diff
@@ -0,0 +1,419 @@
+import chromadb.utils.embedding_functions as ef
+
+
+class AsyncClient(SharedSystemClient, AsyncClientAPI):
+    """A client for Chroma. This is the main entrypoint for interacting with Chroma.
+    A client internally stores its tenant and database and proxies calls to a
+    Server API instance of Chroma. It treats the Server API and corresponding System
+    as a singleton, so multiple clients connecting to the same resource will share the
+    same API instance.
+
+    Client implementations should be implement their own API-caching strategies.
+    """
+
+    # An internal admin client for verifying that databases and tenants exist
+    _admin_client: AsyncAdminAPI
+
+    tenant: str = DEFAULT_TENANT
+    database: str = DEFAULT_DATABASE
+
+    _server: AsyncServerAPI
+
+    @classmethod
+    async def create(
+        cls,
+        tenant: str = DEFAULT_TENANT,
+        database: str = DEFAULT_DATABASE,
+        settings: Settings = Settings(),
+    ) -> "AsyncClient":
+        # Create an admin client for verifying that databases and tenants exist
+        self = cls(settings=settings)
+        SharedSystemClient._populate_data_from_system(self._system)
+        self._admin_client = AsyncAdminClient.from_system(self._system)
+        await self._validate_tenant_database(tenant=tenant, database=database)
+
+        self.tenant = tenant
+        self.database = database
+
+        # Get the root system component we want to interact with
+        self._server = self._system.instance(AsyncServerAPI)
+
+        # Submit event for a client start
+        telemetry_client = self._system.instance(ProductTelemetryClient)
+        telemetry_client.capture(ClientStartEvent())
+
+        return self
+
+    @classmethod
+    async def create_from_system(
+        cls,
+        system: System,
+        tenant: str = DEFAULT_TENANT,
+        database: str = DEFAULT_DATABASE,
+    ) -> "AsyncClient":
+        """Create a client from an existing system. This is useful for testing and debugging."""
+        return await AsyncClient.create(tenant, database, system.settings)
+
+    @classmethod
+    @override
+    def from_system(
+        cls,
+        system: System,
+    ) -> "SharedSystemClient":
+        """AsyncClient cannot be created synchronously. Use .create_from_system() instead."""
+        raise NotImplementedError(
+            "AsyncClient cannot be created synchronously. Use .create_from_system() instead."
+        )
+
+    @override
+    async def set_tenant(self, tenant: str, database: str = DEFAULT_DATABASE) -> None:
+        await self._validate_tenant_database(tenant=tenant, database=database)
+        self.tenant = tenant
+        self.database = database
+
+    @override
+    async def set_database(self, database: str) -> None:
+        await self._validate_tenant_database(tenant=self.tenant, database=database)
+        self.database = database
+
+    async def _validate_tenant_database(self, tenant: str, database: str) -> None:
```

Comment:
it hurts to have to duplicate the error handling here :(

Reply:
yeah unfortunate
I think trying to dry it would make it pretty messy since there's async/sync calls